redis/mutation_utils.cc

```diff
@@ -73,16 +98,73 @@ mutation make_mutation(service::storage_proxy& proxy, const redis_options& optio
     const column_definition& column = *schema->get_column_definition(redis::DATA_COLUMN_NAME);
     auto pkey = partition_key::from_single_value(*schema, key);
     auto m = mutation(schema, std::move(pkey));
-    auto cell = make_cell(schema, *(column.type.get()), data, ttl);
+    auto cell = make_cell(schema, *(column.type.get()), data, api::new_timestamp(), ttl);
     m.set_clustered_cell(clustering_key::make_empty(), column, std::move(cell));
     return m;
 }
 
-future<> write_strings(service::storage_proxy& proxy, redis::redis_options& options, bytes&& key, bytes&& data, long ttl, service_permit permit) {
+future<> write_strings(service::storage_proxy& proxy, redis::redis_options& options, bytes key, bytes data, long ttl, service_permit permit) {
+    // construct cas_request
     db::timeout_clock::time_point timeout = db::timeout_clock::now() + options.get_write_timeout();
-    auto m = make_mutation(proxy, options, std::move(key), std::move(data), ttl);
+    auto schema = get_schema(proxy, options.get_keyspace_name(), redis::STRINGs);
+    const column_definition& column = *schema->get_column_definition(redis::DATA_COLUMN_NAME);
+    auto pkey = partition_key::from_single_value(*schema, key);
+    auto dk = dht::decorate_key(*schema, std::move(pkey));
+    auto partition_range = dht::partition_range::make_singular(dk);
+    dht::partition_range_vector partition_ranges;
+    partition_ranges.emplace_back(std::move(partition_range));
     auto write_consistency_level = options.get_write_consistency_level();
-    return proxy.mutate(std::vector<mutation> {std::move(m)}, write_consistency_level, timeout, nullptr, permit);
+    auto ssg = options.get_smp_service_group();
+
+    service::client_state& client_state = options.get_client_state();
+    auto trace_state = tracing::trace_state_ptr();
+    auto desired_shard = service::storage_proxy::cas_shard(*schema, dk.token());
+    if (desired_shard == this_shard_id()) {
+        auto request = seastar::make_shared<write_strings_cas_request>(proxy, schema, std::move(key), std::move(data), ttl);
+        return proxy.cas(schema, request, nullptr, partition_ranges,
+            {timeout, std::move(permit), client_state, trace_state},
+            db::consistency_level::LOCAL_SERIAL, db::consistency_level::LOCAL_QUORUM, timeout, timeout).then([] (bool is_applied) mutable {
+            assert(is_applied); //XXX: handle failed condition
```

Comment:
I think you can simply ignore is_applied and not even assert it. The is_applied doesn't mean if we had an error in the write (if we did, we'll get an exception, not is_applied=false). is_applied=false just means that the apply() function decided its condition on the read data wasn't true so it returned an empty optional<mutation>. In this write_strings() it can never happen, because apply() never returns an empty mutation. So you can assert() without worrying, or even simpler, just ignore it. In Alternator we have the following in a request (BatchWriteItem) which doesn't have any conditions:
```c++
    return proxy.cas(...).discard_result();
    // We discarded cas()'s future value ("is_applied") because BatchWriteItems
    // does not need to support conditional updates.

```